### PR TITLE
Dtype parameterization

### PR DESCRIPTION
**Pull request by gabrieletijunaityte**
## What does this PR do?

- Added dtype parameter per modality, supported through hydra configs.
- Added dtype parameter for prediction head and normalisation layer, to unify model weight dtypes.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?

### CODE REVIEW

File: src/data/base_dataset.py [hunk 1]
```diff
@@ -288,35 +313,75 @@ def pooch_setup(self) -> None:
...
+        dtype, is_bfloat16 = self.resolve_dtype(dtype)
+
+        arr = self.load_npy(filepath, np.dtype(dtype))
+
+        if arr.shape[-2:] != (size, size):
             arr = center_crop_npy(arr, (128, size, size))
+
+        # Nans are 0 across all 128 channels
+        # mask = np.all(arr == 0, axis=0)
+        # arr[mask] = torch.nan
```
Comment (robknapen) [inline src/data/base_dataset.py]: A bit of a design choice I guess. The NaNs will cause errors, the zeros will flow through silently and affect results. Is 0 a valid value for a tessera (or other EO embedding) dimension?

Comment (gabrieletijunaityte) [reply src/data/base_dataset.py]: By default tessera has 0s for NaNs, distinctly that these 0s are for all 128D. 

We were thinking, that for average encoder it would make sense to ignore these 0s, but for now I left these lines commented out.

Comment (vdplasthijs) [reply src/data/base_dataset.py]: If it's not too complicated; would it be an idea to add an arg to datasets on how to handle NaNs for Tessera (either as NaN or as 0). Then CNN encoders can be configured with 0s, and average encoders with nans (if they can use something like `np.nanmean()`). Or as a first step, let's put this option in the dataset and we can deal with encoders later?

**Review by vdplasthijs — COMMENTED** [review]
Shall we parameterise how Tessera Nans are handled via an arg in base dataset?

File: src/data/base_dataset.py (hunk 1 above)
Comment (robknapen) [reply src/data/base_dataset.py]: Yes, good idea!

File: src/data/base_dataset.py [hunk 2]
```diff
@@ -288,35 +313,75 @@ def pooch_setup(self) -> None:
...
         size = self.modalities["aef"]["size"]
-        if im.shape[1] != size:
+        dtype = self.modalities["aef"].get("dtype")
+        dtype, is_bfloat16 = self.resolve_dtype(dtype)
+
+        im = self.load_tiff(filepath, np.dtype(dtype))
+
+        if im.shape[-2:] != (size, size):
             im = center_crop_npy(im, (64, size, size))
-        if np.isinf(im).any():
```
Comment (gabrieletijunaityte) [inline src/data/base_dataset.py:304]: @vdplasthijs Should we also then preserve aef NaNs?

Comment (vdplasthijs) [reply src/data/base_dataset.py:304]: Yes that makes sense to me (similarly parameterised by an arg)! Thanks! If it's a pain just open it as an issue

Comment (gabrieletijunaityte) [reply src/data/base_dataset.py:304]: We didn't actually have any nans in aef, or I couldn’t find it at least, so it was a bit blind fix 😅 but hopefully works

**Comment by gabrieletijunaityte** [comment]
@vdplasthijs @robknapen it's ready for a second round of checking 😅

File: src/data/base_dataset.py [hunk 3]
```diff
@@ -288,35 +313,81 @@ def pooch_setup(self) -> None:
...
-            im = np.clip(im, a_min=-0.5, a_max=0.5)
-        # TODO any normalisation needed
 
-        return torch.tensor(im).float()
+        # Scan for inf values and clip them (in memory)
+        if self.modalities["aef"].get("enable_nans", False):
```
Comment (vdplasthijs) [inline src/data/base_dataset.py:350]: OK looks good with the enable_nans arg. Was that already defined in modalities, I don't see a new definition in the file changes?

Comment (gabrieletijunaityte) [reply src/data/base_dataset.py:350]: with `get` its an optional arg, which will default to False, I could default it to True?

Comment (robknapen) [reply src/data/base_dataset.py:350]: Looks fine to me as well

**Review by vdplasthijs — APPROVED** [review]
